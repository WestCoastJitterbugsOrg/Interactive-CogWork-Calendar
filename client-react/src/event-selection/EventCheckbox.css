.event-checkbox::after {
  content: "";
  position: relative;
  display: block;
  left: 9px;
  top: 5px;
  width: 6px;
  height: 12px;
  border: transparent;
  transform: rotate(45deg);
}

.event-checkbox.checked::after {
  border-bottom: 2px solid white;
  border-right: 2px solid white;
}


/* Event Group */

.eventgroup-checkbox{
  width: 20px;
  height: 20px;
  background-color: #fffaf2;
  border: 2px solid #1D1D1B;
  border-radius: 6px;
}

.eventgroup-checkbox::after {
  content: "";
  position: relative;
  display: block;
  border: transparent;
}

.eventgroup-checkbox.all::after {
  left: 7px;
  top: -3px;
  width: 8px;
  height: 18px;
  transform: rotate(50deg);
  border-bottom: 3px solid #1D1D1B;
  border-right: 3px solid #1D1D1B;
  border-bottom-right-radius: 3px;
}

.eventgroup-checkbox.some::after {
  top: 2px;
  left: 2px;
  width: 12px;
  height: 12px;
  background-color: #1D1D1B;
  border-radius: 4px;
}
