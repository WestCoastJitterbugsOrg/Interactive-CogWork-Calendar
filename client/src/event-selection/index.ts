import EventGroup from "./event-group";
import EventItem from "./event";
import ToggleAllButtons from "./ToggleAllButtons";

export { EventGroup };
export { EventItem };
export { ToggleAllButtons };
