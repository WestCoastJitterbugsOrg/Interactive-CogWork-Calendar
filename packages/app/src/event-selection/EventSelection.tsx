import { Button } from '../shared/Button';
import { stateContext } from '../state';
import { EventGroup } from './EventGroup';
import style from './EventSelection.module.scss';
import { useCallback, useContext } from 'react';

type Props = {
	isLoading: boolean;
	setCheckedEvents: (events: string[]) => void;
	setEventModal: (eventId: string) => void;
};

export function EventSelection(props: Props) {
	const { events, categories, checkedEvents } = useContext(stateContext);

	const setEventCheckboxState = useCallback(
		(eventId: string, check: boolean) => {
			const checkedEventsSet = new Set(checkedEvents);
			if (check) {
				checkedEventsSet.add(eventId);
			} else {
				checkedEventsSet.delete(eventId);
			}
			props.setCheckedEvents([...checkedEventsSet]);
		},
		[props, checkedEvents],
	);

	const setGroupCheckboxState = useCallback(
		(categoryId: string, state: boolean | 'mixed') => {
			const catEventIds = events
				.filter((e) => e.category === categoryId)
				.map((e) => e.id);

			// Start by unchecking all events in the category
			const newCheckedEvents = checkedEvents.filter(
				(eventId) => !catEventIds.includes(eventId),
			);

			if (state === true) {
				// If the tri-state-checkbox should become checked, add all events in the category
				newCheckedEvents.push(...catEventIds);
			}

			props.setCheckedEvents(newCheckedEvents);
		},
		[props, events, checkedEvents],
	);

	return (
		<>
			<div className={style.actionButtons}>
				<Button
					onClick={() => {
						props.setCheckedEvents(events.map((event) => event.id));
					}}
				>
					Markera allt
				</Button>
				<Button
					onClick={() => {
						props.setCheckedEvents([]);
					}}
				>
					Avmarkera allt
				</Button>
			</div>
			<div className={style.eventGroupList} role="list">
				{categories.map((categoryId) => (
					<EventGroup
						key={categoryId}
						category={categoryId}
						setGroupCheckboxState={(state) => {
							setGroupCheckboxState(categoryId, state);
						}}
						setEventCheckboxstate={setEventCheckboxState}
						setEventModal={props.setEventModal}
					/>
				))}

				{props.isLoading && (
					<div className={style.loader}>
						<div className={style.loaderBar}></div>
					</div>
				)}
			</div>
		</>
	);
}
