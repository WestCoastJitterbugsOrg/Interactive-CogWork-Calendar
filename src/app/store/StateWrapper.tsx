import { canStoreSelection } from '../services/cookies';
import StateContext, { CategoryStore } from './model';
import { createContext, useEffect, useState } from 'react';
import type { WCJ } from 'types';
import type { MaybeArray } from 'types/utils';

export const stateContext = createContext<StateContext>({
	categories: {},
	events: {},
	eventModal: undefined,
});

export type CategoryMap = Readonly<Record<string, CategoryStore>>;
export type EventMap = Readonly<Record<string, Readonly<WCJ.Event>>>;

type Props = {
	children: MaybeArray<JSX.Element>;
	categories: CategoryMap;
	events: EventMap;
	eventModal?: string;
};

export function StateWrapper(props: Props) {
	const [categories, setCategories] = useState<Record<string, CategoryStore>>(
		props.categories
	);
	const [events, setEvents] = useState<EventMap>(props.events);
	const [eventModal, setEventModal] = useState<string | undefined>(
		props.eventModal
	);

	useEffect(() => {
		if (canStoreSelection()) {
			const uncheckedEvents = Object.values(events)
				.filter((x) => !x.showInCalendar)
				.map((x) => x.id);

			localStorage.setItem('uncheckedEvents', JSON.stringify(uncheckedEvents));
		}
	}, [events]);

	return (
		<stateContext.Provider
			value={{
				categories,
				setCategories,
				events,
				setEvents,
				eventModal,
				setEventModal,
			}}
		>
			{props.children}
		</stateContext.Provider>
	);
}
