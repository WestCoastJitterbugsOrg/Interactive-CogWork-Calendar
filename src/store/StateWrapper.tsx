import { createContext, Dispatch, useReducer } from "react";
import { EventActions } from "./event-actions";
import EventStore from "./model";
import eventReducer from "./reducers";

export interface StateContext {
  state: EventStore;
  dispatch: Dispatch<EventActions>;
}
const emptyContext: EventStore = {
  categories: { byId: {}, allIds: [] },
  events: { byId: {}, allIds: [] },
  eventModal: false,
};

export const StateContext = createContext<StateContext>({
  state: emptyContext,
  dispatch: () => null,
});

interface Props {
  children: JSX.Element[];
  initialContext: EventStore;
}

function StateWrapper(props: Props) {
  const [state, dispatch] = useReducer(eventReducer, props.initialContext);

  return (
    <StateContext.Provider value={{ state, dispatch }}>
      {props.children}
    </StateContext.Provider>
  );
}

export default StateWrapper;
