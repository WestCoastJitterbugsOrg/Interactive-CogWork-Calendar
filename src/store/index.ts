import EventStore from "./model";
import eventReducer from "./reducers";

export * from "./hooks";
export * from "./reducers";

export default eventReducer;
export { EventStore };
