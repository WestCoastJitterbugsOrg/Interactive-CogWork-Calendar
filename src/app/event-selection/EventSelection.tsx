import { Button } from '../shared/Button';
import { stateContext } from '../store/StateWrapper';
import WCJ from '../types/wcj';
import { EventGroup } from './EventGroup';
import style from './EventSelection.module.scss';
import { useContext } from 'react';

export function EventSelection() {
	const { categories, setEvents } = useContext(stateContext);

	const select = (show: boolean) => {
		setEvents?.((events) => {
			const newEvents: Record<string, WCJ.Event> = {};
			for (const eventId in events) {
				newEvents[eventId] = {
					...events[eventId],
					showInCalendar: show,
				};
			}
			return newEvents;
		});
	};

	return (
		<>
			<div className={style.actionButtons}>
				<Button onClick={() => select(true)}>Select all</Button>
				<Button onClick={() => select(false)}>Deselect all</Button>
			</div>
			<div
				className={style.eventGroupList}
				data-testid="event-selection-groups"
				role="list"
			>
				{Object.keys(categories).map((categoryId) => (
					<EventGroup key={categoryId} category={categoryId} />
				))}
			</div>
		</>
	);
}
