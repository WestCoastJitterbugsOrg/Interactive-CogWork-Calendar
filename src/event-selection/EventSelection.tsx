import { Button } from "@app/shared/Buttons/Button";
import { StateContext } from "@app/store/StateWrapper";
import { useContext } from "react";
import { EventGroup } from "./EventGroup";

export function EventSelection() {
  const { categories } = useContext(StateContext);
  const { setEvents } = useContext(StateContext) ?? {};

  const select = (show: boolean) =>
    setEvents?.((events) => {
      const newEvents: Record<string, Wcj.Event> = {};

      for (const eventId in events) {
        newEvents[eventId] = {
          ...events[eventId],
          showInCalendar: show,
        };
      }

      return newEvents;
    });
  return (
    <>
      <div className="flex h-16 flex-shrink-0 flex-row items-center justify-center space-x-4 bg-dark">
        <Button onClick={() => select(true)}>Select all</Button>
        <Button onClick={() => select(false)}>Deselect all</Button>
      </div>
      <div
        className="w-full flex-grow divide-y divide-secondary-alt overflow-auto bg-light"
        data-testid="event-selection-groups"
        role="list"
      >
        {Object.keys(categories).map((categoryId) => (
          <EventGroup key={categoryId} category={categoryId} />
        ))}
      </div>
    </>
  );
}
