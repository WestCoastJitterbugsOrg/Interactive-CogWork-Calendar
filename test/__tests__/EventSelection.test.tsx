import { EventSelection } from "@app/event-selection/EventSelection";
import { StateWrapper } from "@app/store/StateWrapper";
import "@testing-library/jest-dom";
import { act, render, RenderResult } from "@testing-library/react";
import { mockStore } from "../__mocks__/stateContext";

it("EventSelection Snapshot", () => {
  const { baseElement } = render(
    <StateWrapper categories={mockStore.categories} events={mockStore.events}>
      <EventSelection />
    </StateWrapper>
  );
  expect(baseElement).toMatchSnapshot();
});

it("Events are selected at start", async () => {
  const renderResult = render(
    <StateWrapper categories={mockStore.categories} events={mockStore.events}>
      <EventSelection />
    </StateWrapper>
  );

  const [groupInput, eventInputs] = await getEventGroupInputs(renderResult);
  expect(groupInput).toBeChecked();
  eventInputs.forEach((eventInput) => expect(eventInput).toBeChecked());
});

it("Clicking deselect all deselects all events and groups", async () => {
  const renderResult = render(
    <StateWrapper categories={mockStore.categories} events={mockStore.events}>
      <EventSelection />
    </StateWrapper>
  );
  const deselectAllBtn = await renderResult.findByText("Deselect all");

  await act(() => deselectAllBtn.click());

  const [groupInput, eventInputs] = await getEventGroupInputs(renderResult);
  expect(groupInput).not.toBeChecked();
  eventInputs.forEach((eventInput) => expect(eventInput).not.toBeChecked());
});

async function getEventGroupInputs(
  renderResult: RenderResult,
  groupSelector = (collection: HTMLCollection) => collection[0]
) {
  const eventGroup = groupSelector(
    (await renderResult.findByTestId("event-selection-groups")).children
  ).children;

  const group = eventGroup[0];
  const eventsWrapper = eventGroup[1];

  const groupInput = group.querySelector("[role='checkbox']");
  const eventInputs = Array.from(
    eventsWrapper.children[0].querySelectorAll("[role='checkbox']")
  );

  return [groupInput, eventInputs] as const;
}
