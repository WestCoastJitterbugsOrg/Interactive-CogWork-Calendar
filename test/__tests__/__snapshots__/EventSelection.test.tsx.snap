// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EventSelection Snapshot 1`] = `
<body>
  <div>
    <div
      class="flex h-16 flex-shrink-0 flex-row items-center justify-center space-x-4 bg-dark"
    >
      <button
        class="button 
    "
        type="button"
      >
        Select all
      </button>
      <button
        class="button 
    "
        type="button"
      >
        Deselect all
      </button>
    </div>
    <div
      class="w-full flex-grow divide-y divide-secondary-alt overflow-auto bg-light"
      data-testid="event-selection-groups"
      role="list"
    >
      <div
        class="bg-light"
        role="group"
      >
        <div
          class="flex min-h-[32px] cursor-pointer flex-row items-center bg-secondary p-2 font-bold text-white"
        >
          <div
            class="flex flex-grow items-center"
          >
            <img
              class="h-4 w-4 flex-shrink-0 transform transition duration-200 "
              src="test-file-stub"
            />
            <span
              class="ml-2"
            >
              Lindy Hop
            </span>
          </div>
          <div
            class="flex"
          >
            <img
              aria-checked="true"
              class="bg-white"
              data-testid="group-checkbox"
              height="16"
              role="checkbox"
              src="test-file-stub"
              width="16"
            />
          </div>
        </div>
        <div
          class="overflow-hidden bg-light max-h-0"
        >
          <div
            class="m-2 flex min-h-[32px] items-center"
            role="listitem"
          >
            <img
              class="mr-2 block h-4 w-4 flex-none cursor-pointer text-black opacity-50 hover:opacity-100"
              data-testid="info-button"
              src="test-file-stub"
            />
            <div
              class="flex flex-grow items-center"
              data-testid="event-item"
            >
              <span
                class="flex-grow cursor-pointer pr-2 leading-tight"
              >
                Grundkurs Lindy Hop
              </span>
              <span
                aria-checked="true"
                aria-readonly="true"
                class="cursor-pointer"
                data-testid="event-checkbox"
                role="checkbox"
              >
                <img
                  height="16"
                  src="test-file-stub"
                  width="16"
                />
              </span>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</body>
`;
